include/psa/crypto.h

```diff
@@ -3750,6 +3760,83 @@ psa_status_t psa_key_derivation_output_key(
     psa_key_derivation_operation_t *operation,
     mbedtls_svc_key_id_t *key);
 
+/** Derive a key from an ongoing key derivation operation with a custom method.
+ *
+ * See the description of psa_key_derivation_out_key() for the operation of
+ * this function with the default method.
+ * Mbed TLS currently does not currently support any non-default methods.
+ *
+ * \note This function is experimental and may change in future minor
+ *       versions of Mbed TLS.
+ *
+ * \param[in] attributes    The attributes for the new key.
+ *                          If the key type to be created is
+ *                          #PSA_KEY_TYPE_PASSWORD_HASH then the algorithm in
+ *                          the policy must be the same as in the current
+ *                          operation.
+ * \param[in,out] operation The key derivation operation object to read from.
+ * \param[in] method        Customization parameters for the key generation.
+ *                          When this is #PSA_KEY_GENERATION_METHOD_INIT
+ *                          with \p method_length =
+ *                          `sizeof(psa_key_generation_method_t)`,
+ *                          this function is equivalent to
+ *                          psa_key_derivation_output_key().
+ *                          Mbed TLS currently only supports the default
+ *                          method, i.e. #PSA_KEY_GENERATION_METHOD_INIT,
+ *                          for all key types.
+ * \param method_length     Length of \p method in bytes.
```

Comment:
The function takes `method_length` as an input. There are alternative ways to handle variable-length structured input:

* `method` could be a structure that can have embedded pointers. That's how domain parameters work. But that requires `malloc` in the client side of client-server implementations and we want to avoid that (TF-M in particular can't handle it).
* `method` could have the length embedded in it somehow. But in client-server implementations, that would require the client to parse the content of `method`.

These alternative ways are not friendly to client-server implementations. In contrast, with `method` not containing embedded pointers and `method_length` parsed separately, the client only needs to know that it must transmit `method_length` bytes to the server, it doesn't need to understand anything about the content.

Reply:
Thanks for the explanation. This anticipated a question I had while reviewing this change :)